### PR TITLE
HTTP client span clarification and establishing HTTP connection spec

### LINKED ISSUE #3155: Establishing HTTP connection spec
**What are you trying to achieve?**

I started working on implementing the `http.resend_count` spec in the Java instrumentations, and I ran into an interesting issue with the current spec.

Currently (that is, without the resends), in pretty much all Java HTTP client instrumentations, we are creating HTTP client spans for the outermost operation, usually the top-level interface exposed by a particular HTTP client library. This means that there is only one HTTP span created for the entire operation; if any resends occur during the call they'll go unnoticed, and the child `SERVER` spans will all receive the same parent span id (which is why we need to modify/rewrite all HTTP client instrumentations we have, they're all too "shallow"). The HTTP span also includes the "establishing the connection" phase (in most cases; I'll mention the exceptions to that later); in case a connection error occurs (closed port, server not responding at all, etc) the HTTP span will end with an error, and the telemetry about the thrown exception will be recorded.

This changes with the resend counter spec. As I understand it, we must now create a span for every attempt to actually send an HTTP request over the wire (and bump the counter if there's a retry). Most HTTP clients that I've read through separate the "establishing the connection" phase from the "sending the request" phase pretty cleanly; which means that whenever a DNS resolution/connection/TLS handshake error occurs, the actual HTTP instrumentation won't catch that - because it runs later than that.

I find this to be somewhat problematic, since usually you'd want to know if this kind of situations occur with your application.

There are several low-level HTTP clients that completely separate these two phases - where any details about the HTTP request are simply not accessible at the connection phase (mostly Netty and Netty-based HTTP clients). For these, we already had to introduce something that'll capture the connection level details, which we've called a [`CONNECT`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ConnectionSpanTest.groovy#L116-L126) span. Note that this already exists in some of our instrumentations, and predates the resend spec.

Now, to solve the problem with the missing connection-level telemetry we could do one of the following things:

* In the top-level HTTP client interface: if there was an exception thrown, and no attempts to send an HTTP request have been made, create a "fake" HTTP span just for the connection error. (Note that this is impossible to implement in the Netty-based clients that I've briefly mentioned above)
* Introduce the `CONNECT` span to the HTTP client spec: create a `CONNECT` span whenever a connection is established (actual connection; taking a connection from a pool should not emit a span), regardless of whether it was successful or not. (We've had customers ask us to implement spans that capture low-level connection details; this kind of information would be generally useful, at least for some people).

**Additional context.**

* HTTP resend PR: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7652
* Some examples of the existing `CONNECT` spans in the javaagent (tests): https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ConnectionSpanTest.groovy#L116-L126, https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/reactor/reactor-netty/reactor-netty-1.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ReactorNettyConnectionSpanTest.java#L99-L107

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Fixes #3155

Ping @trask @lmolkova 

I've added both the additional clarifications about the "low-level" vs "high-level" HTTP span, and the connection span spec. The connection span spec is derived from the current behavior of some of the Java HTTP client instrumentations.

I've made several clauses stronger (SHOULD changed to MUST) to show how I think the two ways of instrumenting HTTP clients should work in principle - consider this a topic for discussion.

### CODE REVIEW

File: specification/trace/semantic_conventions/http.md [hunk 1]
```diff
@@ -428,3 +484,47 @@ GET /hello - 200 (CLIENT, trace=t2, span=s1, http.resend_count=1)
...
+```
+
+Example of establishing an HTTP connection with no trace started upfront:
+
+```
+CONNECT www.example.com:80 (CLIENT, trace=t1, span=s2)
+
+GET /hello - 200 (CLIENT, trace=t1, span=s3)
+ |
+ --- server (SERVER, trace=t1, span=s4)
```
Comment (trask) [inline specification/trace/semantic_conventions/http.md:509]: just checking if I understood this right:

```suggestion
CONNECT www.example.com:80 (CLIENT, trace=t1, span=s1)

GET /hello - 200 (CLIENT, trace=t2, span=s1)
 |
 --- server (SERVER, trace=t2, span=s2)
```

File: specification/trace/semantic_conventions/http.md [hunk 2]
```diff
@@ -428,3 +484,47 @@ GET /hello - 200 (CLIENT, trace=t2, span=s1, http.resend_count=1)
...
+```
+
+Example of an error occurring during establishing an HTTP connection with no trace started upfront:
+
+```
+CONNECT www.example.com:80 (CLIENT, trace=t1, span=s2)
+
+GET /hello - 200 (CLIENT, trace=t1, span=s3)
+ |
+ --- server (SERVER, trace=t1, span=s4)
```
Comment (trask) [inline specification/trace/semantic_conventions/http.md:529]:
```suggestion
CONNECT www.example.com:80 (CLIENT, trace=t1, span=s1)

GET /hello - 200 (CLIENT, trace=t2, span=s1)
 |
 --- server (SERVER, trace=t2, span=s2)
```

File: specification/trace/semantic_conventions/http.md (hunk 1 above)
Comment (mateuszrzeszutek) [reply specification/trace/semantic_conventions/http.md:509]: Ah yes, copy paste mistakes 🙈 
Thanks for catching that!

File: specification/trace/semantic_conventions/http.md [hunk 3]
```diff
@@ -129,12 +132,20 @@ Users MAY explicitly configure instrumentations to capture them even though it i
...
 
-This span type represents an outbound HTTP request.
+This span type represents an outbound HTTP request. There are two ways this can be achieved in an instrumentation:
 
-For an HTTP client span, `SpanKind` MUST be `Client`.
+1. Instrumentations SHOULD create an HTTP span for each attempt to send an HTTP request over the wire.
+   In case the request is resent, the resend attempts MUST follow the [HTTP resend spec](#http-request-retries-and-redirects).
+   Instrumentations MUST NOT emit a "logical" (encompassing) HTTP span.
+   Since establishing an HTTP connection happens before the client attempts to send an HTTP requests,
+   instrumentations SHOULD emit a [CONNECT span](#establishing-an-http-connection) whenever a new HTTP connection is made.
```
Comment (lmolkova) [inline specification/trace/semantic_conventions/http.md:141]: I think that connect span does not cover all cases. If we follow this approach,  to catch other problems, we'd need DNS span and read-stream span and whatnot. I think they can be useful, but verbose and don't solve all problems.

Moreover, in scope of an encompassing span, there could be operations that don't involve any physical HTTP request. E.g.  when circuit-breaking or client-side throttling is used.


I suggest an alternative approach:
1. if instrumentation can create HTTP spans for each attempt, it SHOULD. If it can also instrument logical encompassing operation, it MAY create it, but it's not covered in HTTP semconv v1 (and may be covered in v1.1+). 
If instrumented HTTP client includes DNS and establishing connection into the first try, outer operation does not help much. When these things happen before the instrumentation point, instrumentation SHOULD add outer span too. 

2. if it can't, it SHOULD create HTTP span for topmost operation (as in the p2 there)

With this approach, we can also add CONNECT, DNS, and any other related events and exceptions into the outer operation. 
I think the presence of the outer span should be configurable. E.g. if I use a well-known retry library or client SDK, they should provide outer operation instead of HTTP client instrumentation.

Comment (lmolkova) [reply specification/trace/semantic_conventions/http.md:141]: Example (with a couple of other possible error  points)

image

Comment (mateuszrzeszutek) [reply specification/trace/semantic_conventions/http.md:141]: > I think that connect span does not cover all cases. If we follow this approach, to catch other problems, we'd need DNS span and read-stream span and whatnot. I think they can be useful, but verbose and don't solve all problems.

Yeah, that's true.

> 1. if instrumentation can create HTTP spans for each attempt, it SHOULD. If it can also instrument logical encompassing operation, it MAY create it, but it's not covered in HTTP semconv v1 (and may be covered in v1.1+). It's a different span, it does not have HTTP client semantics

I like the idea of the encompassing span that serves as a sort of catch-all net. It worries me that we're not defining it at all though -- if I were to instrument over-the-wire HTTP requests (and the spec says I SHOULD if it's doable) then I would possibly limit the observability of our instrumentations (e.g. all the network/implementation related events/errors); and since the encompassing span is not really well defined I probably couldn't implement it in a stable instrumentation.

WDYT about emitting both the top-most HTTP span and the over-the-wire HTTP spans? Perhaps with an attribute/annotation explaining that it's a "logical" HTTP span.

> 3. if it can't, it SHOULD create HTTP span for topmost operation (as in the p2 there)

:+1:

> With this approach, we can also add CONNECT, DNS, and any other related events and exceptions into the outer operation.

👍

Comment (lmolkova) [reply specification/trace/semantic_conventions/http.md:141]: I agree that we should somewhat define the outer span, but I don't know if it has to follow any semantic conventions and if it needs to be spec-ed out.

E.g. we can say:

(when applicable) instrument outer logical operation by creating an `INTERNAL`(?) span and record connection errors and events that happen during HTTP request processing before, after, or in between physical HTTP attempts. If no additional information about this logical operation is available, this span SHOULD have "Logical HTTP {VERB}" (?) name or the name of http Client class and method called by user such as `HttpClient.send`. 

If this span has HTTP semantics, it will appear in the results of all queries that match span name or attributes and would skew the results of such queries. 
It's a common problem for any other network instrumentation, and essentially users may want to create such spans themselves to track the overall success/duration of logical operations, so my guess is that we don't need it to have any specific semantics beyond what general otel spec defines.

**Comment by Oberon00** [comment]
Since we are trying to stabilize these conventions: I think this needs to be added nicely separated as an (initially) experimental add-on spec.

File: specification/trace/semantic_conventions/http.md (hunk 3 above)
Comment (Oberon00) [inline specification/trace/semantic_conventions/http.md:141]: I think this name is confusing, since there is also the HTTP verb "CONNECT", which presumably is unrelated.

File: specification/trace/semantic_conventions/http.md [hunk 4]
```diff
@@ -169,18 +180,63 @@ Following attributes MUST be provided **at span creation time** (when provided a
...
 Note that in some cases host and port identifiers in the `Host` header might be different from the `net.peer.name` and `net.peer.port`, in this case instrumentation MAY populate `Host` header on `http.request.header.host` attribute even if it's not enabled by user.
 
+### Establishing an HTTP connection
+
+If an instrumented HTTP client library allows capturing connection level information, instrumentations SHOULD emit a CONNECT span whenever a new HTTP connection is made.
+A CONNECT span contains information that might not get captured by an HTTP client span.
+In particular, if the connection was not established due to some network failure, the HTTP client might return with an error before any HTTP client span is emitted.
+
+Spans emitted when an HTTP connection is established should simply be named `CONNECT`.
+CONNECT spans are not HTTP spans; they do not represent an outgoing HTTP requests and SHOULD NOT contain any `http.*` attributes.
```
Comment (Oberon00) [inline specification/trace/semantic_conventions/http.md:190]: Then, shouldn't we define such a span in some more general network semantic conventions?

File: specification/trace/semantic_conventions/http.md (hunk 3 above)
Comment (mateuszrzeszutek) [reply specification/trace/semantic_conventions/http.md:141]: > * (when applicable...) instrumentation SHOULD create outer logical `INTERNAL`(?) span and record connection errors and events that happen before, after, or in between physical HTTP attempts. If no additional information about this logical operation is available, this span SHOULD have "Logical HTTP {VERB}" (?) name or the name of http client class and method called by user such as `HttpClient.send`.

Hmm, I suppose we could use the `code` semconv for that outer span. At least they would provide some attributes for sampling.

> If this span has HTTP semantics, it will appear in the results of all queries that match span name or attributes and would skew the results of such queries.

👍 

> For example, if user puts `@WithSpan` around their function that prepares request, sends and retries it, and parses the response, it's good enough. And even though we might want to add more stuff on this span, we don't have to define it right away. Correct?

Yeah, that sounds correct.

So, to sum up: "low-level" HTTP instrumentations SHOULD/MAY emit outer `INTERNAL` span with `code` semantics, enabled by default, users can opt-out per instrumentation lib.

**Comment by mateuszrzeszutek** [comment]
Per HTTP SIG meeting: I'll split this PR into 2 and implement @lmolkova 's idea about the encompassing span (instead the CONNECT span)
